File: src/lib/hardware-frames-context.ts
import { bindings } from './binding.js';
import { HardwareDeviceContext } from './hardware-device-context.js';

import type { AVHWFrameTransferDirection, AVPixelFormat } from './constants.js';
import type { Frame } from './frame.js';
import type { NativeHardwareFramesContext, NativeWrapper } from './native-types.js';

/**
 * FFmpeg Hardware Frames Context - Low Level API
 *
 * Direct mapping to FFmpeg's AVHWFramesContext.
 * Manages hardware frame pools and format conversions.
 *
 * @example
 * ```typescript
 * // Create hardware frames context
 * const device = new HardwareDeviceContext();
 * device.create(AV_HWDEVICE_TYPE_CUDA, null, null);
 *
 * const frames = new HardwareFramesContext();
 * frames.alloc(device);
 *
 * // Configure frame parameters
 * frames.format = AV_PIX_FMT_CUDA;     // Hardware format
 * frames.swFormat = AV_PIX_FMT_NV12;   // Software format
 * frames.width = 1920;
 * frames.height = 1080;
 * frames.initialPoolSize = 10;
 *
 * // Initialize the context
 * const ret = frames.init();
 * if (ret < 0) throw new FFmpegError(ret);
 *
 * // Allocate hardware frame
 * const hwFrame = new Frame();
 * hwFrame.alloc();
 * const ret2 = frames.getBuffer(hwFrame, 0);
 * if (ret2 < 0) throw new FFmpegError(ret2);
 *
 * // Transfer between hardware and software
 * const swFrame = new Frame();
 * swFrame.alloc();
 * swFrame.width = 1920;
 * swFrame.height = 1080;
 * swFrame.format = AV_PIX_FMT_NV12;
 * swFrame.getBuffer();
 *
 * // Download from hardware
 * frames.transferData(swFrame, hwFrame, 0);
 *
 * // Upload to hardware
 * frames.transferData(hwFrame, swFrame, 0);
 *
 * // Cleanup
 * hwFrame.free();
 * swFrame.free();
 * frames.free();
 * device.free();
 * ```
 */
export class HardwareFramesContext implements Disposable, NativeWrapper<NativeHardwareFramesContext> {
  private native: NativeHardwareFramesContext;

  // Constructor
  /**
   * Create a new hardware frames context.
   *
   * The context is uninitialized - you must call alloc() before use.
   * Direct wrapper around AVHWFramesContext.
   *
   * @example
   * ```typescript
   * const frames = new HardwareFramesContext();
   * frames.alloc(device);
   * // Configure parameters
   * frames.init();
   * ```
   */
  constructor() {
    this.native = new bindings.HardwareFramesContext();
  }

  // Getter/Setter Properties

  /**
   * Hardware pixel format.
   *
   * Direct mapping to AVHWFramesContext->format
   *
   * The pixel format identifying the underlying HW surface type.
   */
  get format(): AVPixelFormat {
    return this.native.format;
  }

  set format(value: AVPixelFormat) {
    this.native.format = value;
  }

  /**
   * Software pixel format.
   *
   * Direct mapping to AVHWFramesContext->sw_format
   *
   * The pixel format identifying the actual data layout of the hardware frames.
   */
  get swFormat(): AVPixelFormat {
    return this.native.swFormat;
  }

  set swFormat(value: AVPixelFormat) {
    this.native.swFormat = value;
  }

  /**
   * Frame width.
   *
   * Direct mapping to AVHWFramesContext->width
   *
   * The allocated dimensions of the frames in this pool.
   */
  get width(): number {
    return this.native.width;
  }

  set width(value: number) {
    this.native.width = value;
  }

  /**
   * Frame height.
   *
   * Direct mapping to AVHWFramesContext->height
   *
   * The allocated dimensions of the frames in this pool.
   */
  get height(): number {
    return this.native.height;
  }

  set height(value: number) {
    this.native.height = value;
  }

  /**
   * Initial pool size.
   *
   * Direct mapping to AVHWFramesContext->initial_pool_size
   *
   * Initial size of the frame pool. If a device type does not support
   * dynamically resizing the pool, then this is also the maximum pool size.
   */
  get initialPoolSize(): number {
    return this.native.initialPoolSize;
  }

  set initialPoolSize(value: number) {
    this.native.initialPoolSize = value;
  }

  /**
   * Associated device context.
   *
   * Direct mapping to AVHWFramesContext->device_ref
   *
   * @readonly
   */
  get deviceRef(): HardwareDeviceContext | null {
    const native = this.native.deviceRef;
    if (!native) {
      return null;
    }
    // Wrap the native device context
    const device = Object.create(HardwareDeviceContext.prototype) as HardwareDeviceContext;
    (device as any).native = native;
    return device;
  }

  // Public Methods - Lifecycle

  /**
   * Allocate an AVHWFramesContext tied to a given device context.
   *
   * Direct mapping to av_hwframe_ctx_alloc()
   *
   * @param device - Hardware device context
   *
   * @example
   * ```typescript
   * const frames = new HardwareFramesContext();
   * frames.alloc(device);
   * // Set parameters before init()
   * frames.format = AV_PIX_FMT_CUDA;
   * frames.swFormat = AV_PIX_FMT_NV12;
   * frames.width = 1920;
   * frames.height = 1080;
   * ```
   *
   * @throws {Error} If allocation fails
   */
  alloc(device: HardwareDeviceContext): void {
    this.native.alloc(device.getNative());
  }

  /**
   * Finalize the context before use.
   *
   * Direct mapping to av_hwframe_ctx_init()
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - 0: Success
   *   - AVERROR(EINVAL): Invalid parameters
   *   - AVERROR(ENOMEM): Memory allocation failure
   *   - <0: Device-specific errors
   *
   * @example
   * ```typescript
   * const ret = frames.init();
   * if (ret < 0) {
   *   throw new FFmpegError(ret);
   * }
   * ```
   */
  init(): number {
    return this.native.init();
  }

  /**
   * Free the frames context.
   *
   * Unreferences the AVBufferRef.
   *
   * @example
   * ```typescript
   * frames.free();
   * // frames is now invalid and should not be used
   * ```
   */
  free(): void {
    this.native.free();
  }

  // Public Methods - Frame Operations

  /**
   * Allocate a new frame attached to the given AVHWFramesContext.
   *
   * Direct mapping to av_hwframe_get_buffer()
   *
   * @param frame - Frame to allocate (must be allocated but empty)
   * @param flags - Currently unused, should be set to 0
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - 0: Success
   *   - AVERROR(EINVAL): Invalid parameters
   *   - AVERROR(ENOMEM): Memory allocation failure
   *   - <0: Device-specific errors
   *
   * @example
   * ```typescript
   * const frame = new Frame();
   * frame.alloc();
   * const ret = frames.getBuffer(frame, 0);
   * if (ret < 0) {
   *   throw new FFmpegError(ret);
   * }
   * // frame now contains a hardware frame
   * ```
   */
  getBuffer(frame: Frame, flags?: number): number {
    return this.native.getBuffer(frame.getNative(), flags ?? 0);
  }

  /**
   * Copy data to or from a hardware surface.
   *
   * Direct mapping to av_hwframe_transfer_data()
   *
   * @param dst - Destination frame
   * @param src - Source frame
   * @param flags - Currently unused, should be set to 0
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - 0: Success
   *   - AVERROR(ENOSYS): Transfer not supported
   *   - AVERROR(EINVAL): Invalid parameters
   *   - <0: Other errors
   *
   * @example
   * ```typescript
   * // Download from hardware to software
   * const ret = frames.transferData(swFrame, hwFrame, 0);
   * if (ret < 0) {
   *   throw new FFmpegError(ret);
   * }
   *
   * // Upload from software to hardware
   * const ret2 = frames.transferData(hwFrame, swFrame, 0);
   * if (ret2 < 0) {
   *   throw new FFmpegError(ret2);
   * }
   * ```
   */
  transferData(dst: Frame, src: Frame, flags?: number): number {
    return this.native.transferData(dst.getNative(), src.getNative(), flags ?? 0);
  }

  /**
   * Get a list of possible source/target formats.
   *
   * Direct mapping to av_hwframe_transfer_get_formats()
   *
   * @param direction - Transfer direction (AV_HWFRAME_TRANSFER_DIRECTION_*)
   *
   * @returns Array of pixel formats or error code:
   *   - Array: List of supported formats
   *   - <0: AVERROR on failure
   *
   * @example
   * ```typescript
   * // Get formats for downloading from hardware
   * const formats = frames.transferGetFormats(AV_HWFRAME_TRANSFER_DIRECTION_FROM);
   * if (Array.isArray(formats)) {
   *   console.log('Supported download formats:', formats);
   * }
   *
   * // Get formats for uploading to hardware
   * const formats2 = frames.transferGetFormats(AV_HWFRAME_TRANSFER_DIRECTION_TO);
   * if (Array.isArray(formats2)) {
   *   console.log('Supported upload formats:', formats2);
   * }
   * ```
   */
  transferGetFormats(direction: AVHWFrameTransferDirection): AVPixelFormat[] | number {
    return this.native.transferGetFormats(direction);
  }

  /**
   * Map a hardware frame.
   *
   * Direct mapping to av_hwframe_map()
   *
   * This has a number of different possible effects, depending on the format
   * and origin of the src and dst frames. On input, src should be a usable
   * frame with valid hardware/software format and allocated data.
   *
   * @param dst - Destination frame
   * @param src - Source frame to map
   * @param flags - Combination of AV_HWFRAME_MAP_* flags
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - 0: Success
   *   - AVERROR(ENOSYS): Mapping not supported
   *   - AVERROR(EINVAL): Invalid parameters
   *   - <0: Other errors
   *
   * @example
   * ```typescript
   * const mappedFrame = new Frame();
   * mappedFrame.alloc();
   * const ret = frames.map(mappedFrame, hwFrame, AV_HWFRAME_MAP_READ);
   * if (ret < 0) {
   *   throw new FFmpegError(ret);
   * }
   * // mappedFrame now provides CPU access to hwFrame
   * ```
   */
  map(dst: Frame, src: Frame, flags?: number): number {
    return this.native.map(dst.getNative(), src.getNative(), flags ?? 0);
  }

  /**
   * Create a new frame context derived from an existing one.
   *
   * Direct mapping to av_hwframe_ctx_create_derived()
   *
   * @param format - Pixel format for the derived context
   * @param derivedDevice - Device context for the derived frames
   * @param source - Source frames context
   * @param flags - Currently unused, should be 0
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - 0: Success
   *   - AVERROR(ENOSYS): Derivation not supported
   *   - AVERROR(EINVAL): Invalid parameters
   *   - <0: Other errors
   *
   * @example
   * ```typescript
   * const derived = new HardwareFramesContext();
   * const ret = derived.createDerived(
   *   AV_PIX_FMT_NV12,
   *   derivedDevice,
   *   sourceFrames,
   *   0
   * );
   * if (ret < 0) {
   *   throw new FFmpegError(ret);
   * }
   * ```
   */
  createDerived(format: AVPixelFormat, derivedDevice: HardwareDeviceContext, source: HardwareFramesContext, flags?: number): number {
    return this.native.createDerived(format, derivedDevice.getNative(), source.native, flags ?? 0);
  }

  // Internal Methods

  /**
   * Get the native FFmpeg AVHWFramesContext pointer.
   *
   * @internal For use by other wrapper classes
   * @returns The underlying native hardware frames context object
   */
  getNative(): NativeHardwareFramesContext {
    return this.native;
  }

  /**
   * Dispose of the hardware frames context.
   *
   * Implements the Disposable interface for automatic cleanup.
   * Equivalent to calling free().
   *
   * @example
   * ```typescript
   * {
   *   using frames = new HardwareFramesContext();
   *   frames.alloc(device);
   *   frames.init();
   *   // ... use frames
   * } // Automatically freed when leaving scope
   * ```
   */
  [Symbol.dispose](): void {
    this.native[Symbol.dispose]();
  }
}

File: src/bindings/hardware_frames_context.cc
#include "hardware_frames_context.h"
#include "hardware_device_context.h"
#include "frame.h"
#include "error.h"

namespace ffmpeg {

Napi::FunctionReference HardwareFramesContext::constructor;

Napi::Object HardwareFramesContext::Init(Napi::Env env, Napi::Object exports) {
  Napi::Function func = DefineClass(env, "HardwareFramesContext", {
    // Methods - Low Level API
    InstanceMethod<&HardwareFramesContext::Alloc>("alloc"),
    InstanceMethod<&HardwareFramesContext::Init>("init"),
    InstanceMethod<&HardwareFramesContext::GetBuffer>("getBuffer"),
    InstanceMethod<&HardwareFramesContext::TransferData>("transferData"),
    InstanceMethod<&HardwareFramesContext::TransferGetFormats>("transferGetFormats"),
    InstanceMethod<&HardwareFramesContext::Map>("map"),
    InstanceMethod<&HardwareFramesContext::CreateDerived>("createDerived"),
    InstanceMethod<&HardwareFramesContext::Free>("free"),

    // Properties
    InstanceAccessor("format", &HardwareFramesContext::GetFormat, &HardwareFramesContext::SetFormat),
    InstanceAccessor("swFormat", &HardwareFramesContext::GetSwFormat, &HardwareFramesContext::SetSwFormat),
    InstanceAccessor("width", &HardwareFramesContext::GetWidth, &HardwareFramesContext::SetWidth),
    InstanceAccessor("height", &HardwareFramesContext::GetHeight, &HardwareFramesContext::SetHeight),
    InstanceAccessor("initialPoolSize", &HardwareFramesContext::GetInitialPoolSize, &HardwareFramesContext::SetInitialPoolSize),
    InstanceAccessor("deviceRef", &HardwareFramesContext::GetDeviceRef, nullptr),
    
    // Utility
    InstanceMethod(Napi::Symbol::WellKnown(env, "dispose"), &HardwareFramesContext::Dispose),
  });
  
  constructor = Napi::Persistent(func);
  constructor.SuppressDestruct();
  
  exports.Set("HardwareFramesContext", func);
  return exports;
}

// === Lifecycle ===

HardwareFramesContext::HardwareFramesContext(const Napi::CallbackInfo& info) 
  : Napi::ObjectWrap<HardwareFramesContext>(info), unowned_ref_(nullptr) {
  // Constructor does nothing - user must call alloc()
}

HardwareFramesContext::~HardwareFramesContext() {
  // Manual cleanup if not already done
  if (frames_ref_ && !is_freed_) {
    av_buffer_unref(&frames_ref_);
    frames_ref_ = nullptr;
  }
  // RAII handles cleanup
}

// === Static Methods ===

Napi::Value HardwareFramesContext::Wrap(Napi::Env env, AVBufferRef* frames_ref) {
  if (!frames_ref) {
    return env.Null();
  }
  
  Napi::Object obj = constructor.New({});
  HardwareFramesContext* ctx = Napi::ObjectWrap<HardwareFramesContext>::Unwrap(obj);
  ctx->SetUnowned(frames_ref); // We don't own contexts that are wrapped
  return obj;
}

// === Methods ===

Napi::Value HardwareFramesContext::Alloc(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 1 || !info[0].IsObject()) {
    Napi::TypeError::New(env, "Expected HardwareDeviceContext")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (Get()) {
    Napi::Error::New(env, "Frames context already allocated").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  HardwareDeviceContext* device = Napi::ObjectWrap<HardwareDeviceContext>::Unwrap(info[0].As<Napi::Object>());
  if (!device || !device->Get()) {
    Napi::Error::New(env, "Invalid device context").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  AVBufferRef* new_ref = av_hwframe_ctx_alloc(device->Get());
  if (!new_ref) {
    Napi::Error::New(env, "Failed to allocate hardware frames context").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (frames_ref_ && !is_freed_) { av_buffer_unref(&frames_ref_); } frames_ref_ = new_ref; is_freed_ = false;
  return env.Undefined();
}

Napi::Value HardwareFramesContext::Init(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref) {
    Napi::Error::New(env, "Frames context not allocated").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int ret = av_hwframe_ctx_init(ref);
  return Napi::Number::New(env, ret);
}

Napi::Value HardwareFramesContext::GetBuffer(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 1 || !info[0].IsObject()) {
    Napi::TypeError::New(env, "Expected Frame object")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  AVBufferRef* ref = Get();
  if (!ref) {
    Napi::Error::New(env, "Frames context not allocated").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  Frame* frame = Napi::ObjectWrap<Frame>::Unwrap(info[0].As<Napi::Object>());
  if (!frame || !frame->Get()) {
    Napi::Error::New(env, "Invalid frame").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int flags = 0;
  if (info.Length() > 1 && info[1].IsNumber()) {
    flags = info[1].As<Napi::Number>().Int32Value();
  }
  
  int ret = av_hwframe_get_buffer(ref, frame->Get(), flags);
  return Napi::Number::New(env, ret);
}

// TransferData is now implemented in hardware_frames_context_async.cc

Napi::Value HardwareFramesContext::TransferGetFormats(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref) {
    Napi::Error::New(env, "Frames context not allocated").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (info.Length() < 1 || !info[0].IsNumber()) {
    Napi::TypeError::New(env, "Expected direction (AV_HWFRAME_TRANSFER_DIRECTION_*)")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  enum AVHWFrameTransferDirection dir = static_cast<AVHWFrameTransferDirection>(
    info[0].As<Napi::Number>().Int32Value()
  );
  
  enum AVPixelFormat* formats = nullptr;
  int ret = av_hwframe_transfer_get_formats(ref, dir, &formats, 0);
  
  if (ret < 0 || !formats) {
    return Napi::Number::New(env, ret);
  }
  
  // Count formats
  int count = 0;
  while (formats[count] != AV_PIX_FMT_NONE) {
    count++;
  }
  
  // Create array
  Napi::Array arr = Napi::Array::New(env, count);
  for (int i = 0; i < count; i++) {
    arr[i] = Napi::Number::New(env, formats[i]);
  }
  
  av_free(formats);
  return arr;
}

Napi::Value HardwareFramesContext::Map(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 2 || !info[0].IsObject() || !info[1].IsObject()) {
    Napi::TypeError::New(env, "Expected (dst: Frame, src: Frame, flags?: number)")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  Frame* dst = Napi::ObjectWrap<Frame>::Unwrap(info[0].As<Napi::Object>());
  Frame* src = Napi::ObjectWrap<Frame>::Unwrap(info[1].As<Napi::Object>());
  
  if (!dst || !dst->Get() || !src || !src->Get()) {
    Napi::Error::New(env, "Invalid frame(s)").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int flags = 0;
  if (info.Length() > 2 && info[2].IsNumber()) {
    flags = info[2].As<Napi::Number>().Int32Value();
  }
  
  int ret = av_hwframe_map(dst->Get(), src->Get(), flags);
  return Napi::Number::New(env, ret);
}

Napi::Value HardwareFramesContext::CreateDerived(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 3 || !info[0].IsNumber() || !info[1].IsObject() || !info[2].IsObject()) {
    Napi::TypeError::New(env, "Expected (format: number, derivedDevice: HardwareDeviceContext, sourceFrames: HardwareFramesContext, flags?: number)")
        .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  enum AVPixelFormat format = static_cast<AVPixelFormat>(info[0].As<Napi::Number>().Int32Value());
  
  HardwareDeviceContext* derivedDevice = Napi::ObjectWrap<HardwareDeviceContext>::Unwrap(info[1].As<Napi::Object>());
  if (!derivedDevice || !derivedDevice->Get()) {
    Napi::Error::New(env, "Invalid derived device context").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  HardwareFramesContext* src = Napi::ObjectWrap<HardwareFramesContext>::Unwrap(info[2].As<Napi::Object>());
  if (!src || !src->Get()) {
    Napi::Error::New(env, "Invalid source frames context").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int flags = 0;
  if (info.Length() > 3 && info[3].IsNumber()) {
    flags = info[3].As<Napi::Number>().Int32Value();
  }
  
  // Free existing context if any
  if (frames_ref_ && !is_freed_) { av_buffer_unref(&frames_ref_); frames_ref_ = nullptr; is_freed_ = true; }
  unowned_ref_ = nullptr;
  
  AVBufferRef* new_ref = nullptr;
  int ret = av_hwframe_ctx_create_derived(&new_ref, format, derivedDevice->Get(), src->Get(), flags);
  
  if (ret >= 0 && new_ref) {
    if (frames_ref_ && !is_freed_) { av_buffer_unref(&frames_ref_); } frames_ref_ = new_ref; is_freed_ = false;
  }
  
  return Napi::Number::New(env, ret);
}

Napi::Value HardwareFramesContext::Free(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (frames_ref_ && !is_freed_) { av_buffer_unref(&frames_ref_); frames_ref_ = nullptr; is_freed_ = true; }
  unowned_ref_ = nullptr;
  
  return env.Undefined();
}

// === Properties ===

Napi::Value HardwareFramesContext::GetFormat(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Undefined();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  return Napi::Number::New(env, ctx->format);
}

void HardwareFramesContext::SetFormat(const Napi::CallbackInfo& info, const Napi::Value& value) {
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return;
  }
  
  if (!value.IsNumber()) {
    return;
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  ctx->format = static_cast<AVPixelFormat>(value.As<Napi::Number>().Int32Value());
}

Napi::Value HardwareFramesContext::GetSwFormat(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Undefined();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  return Napi::Number::New(env, ctx->sw_format);
}

void HardwareFramesContext::SetSwFormat(const Napi::CallbackInfo& info, const Napi::Value& value) {
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return;
  }
  
  if (!value.IsNumber()) {
    return;
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  ctx->sw_format = static_cast<AVPixelFormat>(value.As<Napi::Number>().Int32Value());
}

Napi::Value HardwareFramesContext::GetWidth(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Undefined();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  return Napi::Number::New(env, ctx->width);
}

void HardwareFramesContext::SetWidth(const Napi::CallbackInfo& info, const Napi::Value& value) {
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return;
  }
  
  if (!value.IsNumber()) {
    return;
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  ctx->width = value.As<Napi::Number>().Int32Value();
}

Napi::Value HardwareFramesContext::GetHeight(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Undefined();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  return Napi::Number::New(env, ctx->height);
}

void HardwareFramesContext::SetHeight(const Napi::CallbackInfo& info, const Napi::Value& value) {
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return;
  }
  
  if (!value.IsNumber()) {
    return;
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  ctx->height = value.As<Napi::Number>().Int32Value();
}

Napi::Value HardwareFramesContext::GetInitialPoolSize(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Undefined();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  return Napi::Number::New(env, ctx->initial_pool_size);
}

void HardwareFramesContext::SetInitialPoolSize(const Napi::CallbackInfo& info, const Napi::Value& value) {
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return;
  }
  
  if (!value.IsNumber()) {
    return;
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  ctx->initial_pool_size = value.As<Napi::Number>().Int32Value();
}

Napi::Value HardwareFramesContext::GetDeviceRef(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  AVBufferRef* ref = Get();
  if (!ref || !ref->data) {
    return env.Null();
  }
  
  AVHWFramesContext* ctx = reinterpret_cast<AVHWFramesContext*>(ref->data);
  if (!ctx->device_ref) {
    return env.Null();
  }
  
  return HardwareDeviceContext::Wrap(env, ctx->device_ref);
}

// === Utility ===

Napi::Value HardwareFramesContext::Dispose(const Napi::CallbackInfo& info) {
  return Free(info);
}

} // namespace ffmpeg